File: app/src/main/java/winnipegbusbackend/FavouriteStopsList.java
package winnipegbusbackend;

import android.util.Xml;

import com.kieran.winnipegbus.HomeScreenActivity;

import org.w3c.dom.Document;
import org.w3c.dom.NodeList;
import org.xmlpull.v1.XmlSerializer;

import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;

public class FavouriteStopsList {
    static String FILES_DIR = HomeScreenActivity.filesDir + "/favourites.xml";
    public static List<FavouriteStop> favouritesList = new ArrayList<>();

    public static void addToFavourites(FavouriteStop favouriteStop) {
        if(!contains(favouriteStop)) {
            favouritesList.add(favouriteStop);
            saveFavouriteStops();
        }
    }

    public static boolean contains(FavouriteStop favouriteStop) {
        for(FavouriteStop fs : favouritesList)
            if(fs.getStopNumber() == favouriteStop.getStopNumber())
                return true;

        return false;
    }

    public static boolean contains(int stopNumber) {
        for(FavouriteStop fs : favouritesList)
            if(fs.getStopNumber() == stopNumber)
                return true;

        return false;
    }

    public static void removeFromFavourites(int stopNumber) {
        favouritesList.remove(getFavouriteStopByStopNumber(stopNumber));
        saveFavouriteStops();
    }

    public static FavouriteStop getFavouriteStopByStopNumber(int stopNumber) {
        for(FavouriteStop fs : favouritesList)
            if(fs.getStopNumber() == stopNumber)
                return fs;
        return null;
    }

   public static void loadFavourites() {
        BusUtilities utilities = new BusUtilities();

        try {
            Document XMLDocument = utilities.getXML(new FileInputStream(FILES_DIR));
            NodeList stopNumbers = XMLDocument.getElementsByTagName(FavouritesNodeTags.STOP_NUMBER.tag);
            NodeList stopNames = XMLDocument.getElementsByTagName(FavouritesNodeTags.STOP_NAME.tag);
            NodeList timesUsed = XMLDocument.getElementsByTagName(FavouritesNodeTags.TIMES_USED.tag);

            for (int r = 0; r < stopNumbers.getLength(); r++)
                addToFavourites(new FavouriteStop(stopNames.item(r).getFirstChild().getNodeValue(), Integer.parseInt(stopNumbers.item(r).getFirstChild().getNodeValue()), Integer.parseInt(timesUsed.item(r).getFirstChild().getNodeValue())));

        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public static boolean saveFavouriteStops() {
        try {
            FileOutputStream fos;
            fos = new FileOutputStream(FILES_DIR);
            XmlSerializer serializer = Xml.newSerializer();
            serializer.setOutput(fos, "UTF-8");
            serializer.startTag("", FavouritesNodeTags.FAVOURITE_STOPS.tag);

            serializer.setFeature("http://xmlpull.org/v1/doc/features.html#indent-output", true);

            for (int i = 0; i < favouritesList.size(); i++) {
                serializer.startTag("", FavouritesNodeTags.FAVOURITE_STOP.tag);
                serializer.startTag("", FavouritesNodeTags.STOP_NUMBER.tag);
                serializer.text(Integer.toString(favouritesList.get(i).getStopNumber()));
                serializer.endTag("", FavouritesNodeTags.STOP_NUMBER.tag);

                serializer.startTag("", FavouritesNodeTags.STOP_NAME.tag);
                serializer.text(favouritesList.get(i).getStopName());
                serializer.endTag("", FavouritesNodeTags.STOP_NAME.tag);

                serializer.startTag("", FavouritesNodeTags.TIMES_USED.tag);
                serializer.text(Integer.toString(favouritesList.get(i).getTimesUsed()));
                serializer.endTag("", FavouritesNodeTags.TIMES_USED.tag);

                serializer.endTag("", FavouritesNodeTags.FAVOURITE_STOP.tag);
            }
            serializer.endTag("", FavouritesNodeTags.FAVOURITE_STOPS.tag);
            serializer.endDocument();
            serializer.flush();
            fos.close();
            return true;
        } catch (IOException e) {
            return false;
        }
    }

    public static List<FavouriteStop> getFavouriteStopsSorted(final int sortTypeId) {
        List<FavouriteStop> favouritesList = FavouriteStopsList.favouritesList;
        List<FavouriteStop> sortedFavouritesList = new ArrayList<>();

        for (FavouriteStop f : favouritesList)
            sortedFavouritesList.add(FavouriteStop.clone(f));

        Collections.sort(sortedFavouritesList, new Comparator<FavouriteStop>() {
            @Override
            public int compare(FavouriteStop stop1, FavouriteStop stop2) {
                if(sortTypeId == FavouritesListSortTypeId.SAVED_INDEX.value)
                    return 0;
                else if(sortTypeId == FavouritesListSortTypeId.STOP_NUBMER_ASC.value)
                    return stop1.getStopNumber() - stop2.getStopNumber();
                else if(sortTypeId == FavouritesListSortTypeId.STOP_NUBMER_DESC.value)
                    return -(stop1.getStopNumber() - stop2.getStopNumber());
                else if(sortTypeId == FavouritesListSortTypeId.FREQUENCY_ASC.value)
                    return (stop1.getTimesUsed() - stop2.getTimesUsed());
                else if(sortTypeId == FavouritesListSortTypeId.FREQUENCY_DESC.value)
                    return -(stop1.getTimesUsed() - stop2.getTimesUsed());
                return 0;
            }
        });

        return sortedFavouritesList;
    }
}

File: app/src/main/java/com/kieran/winnipegbusbackend/agency/winnipegtransit/FavouriteStopsList.kt
package com.kieran.winnipegbusbackend.agency.winnipegtransit

import android.util.Xml
import com.kieran.winnipegbus.activities.BaseActivity
import com.kieran.winnipegbusbackend.common.FavouriteStop
import com.kieran.winnipegbusbackend.common.LoadResult
import com.kieran.winnipegbusbackend.enums.FavouritesListSortType
import org.w3c.dom.Document
import org.w3c.dom.Element
import org.w3c.dom.Node
import java.io.FileInputStream
import java.io.FileOutputStream
import java.io.IOException
import java.io.InputStream
import java.util.*
import javax.xml.parsers.DocumentBuilderFactory

object FavouriteStopsList {
    private val FILES_DIR = BaseActivity.filesDirPath!! + "/favourites.xml"

    private val FAVOURITE_STOPS_TAG = "favouriteStops"
    private val FAVOURITE_STOP_TAG = "favouriteStop"
    private val STOP_NUMBER_TAG = "stopNumber"
    private val STOP_NAME_TAG = "stopName"
    private val TIMES_USED_TAG = "timesUsed"
    private val ALIAS_TAG = "alias"

    private val favouritesList = ArrayList<FavouriteStop>()
    var isLoadNeeded = true
    private val XMLFeature = "http://xmlpull.org/v1/doc/features.html#indent-output"

    operator fun contains(favouriteStop: FavouriteStop): Boolean {
        return contains((favouriteStop.identifier as WinnipegTransitStopIdentifier).stopNumber)
    }

    operator fun contains(stopNumber: Int): Boolean {
        return favouritesList.any { (it.identifier as WinnipegTransitStopIdentifier).stopNumber == stopNumber }
    }

    fun loadFavourites(): Boolean {
        if (favouritesList.isEmpty())
            isLoadNeeded = true
        if (isLoadNeeded) {
            try {
                val XMLDocument = getXML(FileInputStream(FILES_DIR)).result
                val favouriteStops = XMLDocument!!.getElementsByTagName(FAVOURITE_STOP_TAG)

                for (r in 0 until favouriteStops.length) {
                    val curr = favouriteStops.item(r)
                    val stopNumber = Integer.parseInt(getValue(STOP_NUMBER_TAG, curr))
                    val stopName = getValue(STOP_NAME_TAG, favouriteStops.item(r))
                    val timesUsed = Integer.parseInt(getValue(TIMES_USED_TAG, curr))
                    val alias = getValue(ALIAS_TAG, favouriteStops.item(r))

                    val favouriteStop = FavouriteStop(stopName!!, WinnipegTransitStopIdentifier(stopNumber), timesUsed)
                    if (alias != null)
                        favouriteStop.alias = alias

                    if (!contains(favouriteStop))
                        favouritesList.add(favouriteStop)
                }
                isLoadNeeded = false
            } catch (e: Exception) {
                isLoadNeeded = true
            }

        }
        return isLoadNeeded
    }

    fun sort(sortType: FavouritesListSortType) {
        Collections.sort(favouritesList) { stop1, stop2 ->
            when (sortType) {
                FavouritesListSortType.STOP_NUMBER_ASC -> (stop1.identifier as WinnipegTransitStopIdentifier).stopNumber - (stop2.identifier as WinnipegTransitStopIdentifier).stopNumber
                FavouritesListSortType.STOP_NUMBER_DESC -> -((stop1.identifier as WinnipegTransitStopIdentifier).stopNumber - (stop2.identifier as WinnipegTransitStopIdentifier).stopNumber)
                FavouritesListSortType.FREQUENCY_ASC -> stop1.timesUsed - stop2.timesUsed
                FavouritesListSortType.FREQUENCY_DESC -> -(stop1.timesUsed - stop2.timesUsed)
            }
        }
    }

    fun getFavouriteStopsSorted(sortType: FavouritesListSortType, maxItems: Int = -1): List<FavouriteStop> {
        loadFavourites()

        sort(sortType)

        return if(maxItems == -1) {
            favouritesList
        }else {
            favouritesList.take(maxItems)
        }
    }

    operator fun get(position: Int): FavouriteStop {
        return favouritesList[position]
    }

    private fun getValue(tag: String, originalNode: Node): String? {
        return try {
            val node = (originalNode as Element).getElementsByTagName(tag).item(0).firstChild
            node.nodeValue
        } catch (e: Exception) {
            null
        }

    }

    private fun getXML(inputStream: InputStream): LoadResult<Document> {
        return try {
            val dbf = DocumentBuilderFactory.newInstance()
            val db = dbf.newDocumentBuilder()
            val XMLDocument = db.parse(inputStream)

            LoadResult(XMLDocument, null)
        } catch (e: Exception) {
            LoadResult<Document>(null, e)
        }

    }
}
